<a href="https://colab.research.google.com/github/vipulgote1999/machine-learning-udacity/blob/master/svm_udacity_comleted.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
!pip install scikit-learn==0.18


     |████████████████████████████████| 11.4MB 34.8MB/s 
ERROR: fancyimpute 0.4.3 has requirement scikit-learn>=0.19.1, but you'll have scikit-learn 0.18 which is incompatible.
ERROR: imbalanced-learn 0.4.3 has requirement scikit-learn>=0.20, but you'll have scikit-learn 0.18 which is incompatible.
ERROR: yellowbrick 0.9.1 has requirement scikit-learn>=0.20, but you'll have scikit-learn 0.18 which is incompatible.
  Found existing installation: scikit-learn 0.20.3
    Uninstalling scikit-learn-0.20.3:
      Successfully uninstalled scikit-learn-0.20.3


In [0]:
!pip install scipy==0.18.1
!pip install numpy==1.11.2
!pip install nltk==3.2.1

     |████████████████████████████████| 40.3MB 73.4MB/s 
ERROR: cvxpy 1.0.15 has requirement scipy>=1.1.0, but you'll have scipy 0.18.1 which is incompatible.
ERROR: albumentations 0.1.12 has requirement imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.9 which is incompatible.
ERROR: umap-learn 0.3.9 has requirement scipy>=0.19, but you'll have scipy 0.18.1 which is incompatible.
ERROR: mir-eval 0.5 has requirement scipy>=1.0.0, but you'll have scipy 0.18.1 which is incompatible.
ERROR: fastai 0.7.0 has requirement torch<0.4, but you'll have torch 1.1.0 which is incompatible.
ERROR: librosa 0.6.3 has requirement scipy>=1.0.0, but you'll have scipy 0.18.1 which is incompatible.
ERROR: fancyimpute 0.4.3 has requirement scikit-learn>=0.19.1, but you'll have scikit-learn 0.18 which is incompatible.
ERROR: imbalanced-learn 0.4.3 has requirement scikit-learn>=0.20, but you'll have scikit-learn 0.18 which is incompatible.
ERROR: yellowbrick 0.9.1 has requirement scikit-learn>=0.20, but you'll

     |████████████████████████████████| 1.1MB 45.2MB/s 
  Created wheel for nltk: filename=nltk-3.2.1-cp27-none-any.whl size=1316641 sha256=81326b16aec66860e324990cbfc373b32df6c894b350d41555e7586e1285d2ba
  Stored in directory: /root/.cache/pip/wheels/15/09/54/c7baf815a67c5f19ec23752715885b1ce25c9617e8a1c25da3
Successfully built nltk
  Found existing installation: nltk 3.2.5
    Uninstalling nltk-3.2.5:
      Successfully uninstalled nltk-3.2.5


In [0]:
!git clone https://github.com/udacity/ud120-projects.git

Cloning into 'ud120-projects'...
remote: Enumerating objects: 5046, done.
remote: Total 5046 (delta 0), reused 0 (delta 0), pack-reused 5046
Receiving objects: 100% (5046/5046), 19.64 MiB | 3.47 MiB/s, done.
Resolving deltas: 100% (4372/4372), done.


In [0]:
#!/usr/bin/python

import pickle
import cPickle
import numpy

from sklearn import cross_validation
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_selection import SelectPercentile, f_classif



def preprocess(words_file = "/content/ud120-projects/tools/word_data.pkl", authors_file="/content/ud120-projects/tools/email_authors.pkl"):
    """ 
        this function takes a pre-made list of email texts (by default word_data.pkl)
        and the corresponding authors (by default email_authors.pkl) and performs
        a number of preprocessing steps:
            -- splits into training/testing sets (10% testing)
            -- vectorizes into tfidf matrix
            -- selects/keeps most helpful features

        after this, the feaures and labels are put into numpy arrays, which play nice with sklearn functions

        4 objects are returned:
            -- training/testing features
            -- training/testing labels

    """

    ### the words (features) and authors (labels), already largely preprocessed
    ### this preprocessing will be repeated in the text learning mini-project
    authors_file_handler = open(authors_file, "r")
    authors = pickle.load(authors_file_handler)
    authors_file_handler.close()

    words_file_handler = open(words_file, "r")
    word_data = cPickle.load(words_file_handler)
    words_file_handler.close()

    ### test_size is the percentage of events assigned to the test set
    ### (remainder go into training)
    features_train, features_test, labels_train, labels_test = cross_validation.train_test_split(word_data, authors, test_size=0.1, random_state=42)



    ### text vectorization--go from strings to lists of numbers
    vectorizer = TfidfVectorizer(sublinear_tf=True, max_df=0.5,
                                 stop_words='english')
    features_train_transformed = vectorizer.fit_transform(features_train)
    features_test_transformed  = vectorizer.transform(features_test)



    ### feature selection, because text is super high dimensional and 
    ### can be really computationally chewy as a result
    selector = SelectPercentile(f_classif, percentile=10)
    selector.fit(features_train_transformed, labels_train)
    features_train_transformed = selector.transform(features_train_transformed).toarray()
    features_test_transformed  = selector.transform(features_test_transformed).toarray()

    ### info on the data
    print "no. of Chris training emails:", sum(labels_train)
    print "no. of Sara training emails:", len(labels_train)-sum(labels_train)
    
    return features_train_transformed, features_test_transformed, labels_train, labels_test


/usr/local/lib/python2.7/dist-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [0]:


""" 
    This is the code to accompany the Lesson 2 (SVM) mini-project.

    Use a SVM to identify emails from the Enron corpus by their authors:    
    Sara has label 0
    Chris has label 1
"""
    
import sys
from time import time
from sklearn import svm
from sklearn.metrics import accuracy_score

#########################################################
### your code goes here ###

#########################################################

### features_train and features_test are the features for the training
### and testing datasets, respectively
### labels_train and labels_test are the corresponding item labels
features_train, features_test, labels_train, labels_test = preprocess()
model=svm.SVC(kernel='rbf',C=10000)
#features_train = features_train[:len(features_train)/100] 
#labels_train = labels_train[:len(labels_train)/100] 
print(len(labels_train))






/usr/local/lib/python2.7/dist-packages/sklearn/feature_extraction/text.py:1039: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  if hasattr(X, 'dtype') and np.issubdtype(X.dtype, np.float):
/usr/local/lib/python2.7/dist-packages/sklearn/utils/__init__.py:54: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(mask.dtype, np.int):
/usr/local/lib/python2.7/dist-packages/scipy/stats/stats.py:1626: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  return np.add.reduce(sorted[indexer] * weights, axi

no. of Chris training emails: 7936
no. of Sara training emails: 7884
15820


In [0]:
model.fit(features_train,labels_train)
t0=time()

pred=model.predict(features_test)

print "training time:", round(time()-t0, 3), "s"

training time: 17.354 s


In [0]:
s=[]
for i in pred:
  if i==1:
    s.append(i)
print(len(s))

877


no. of Chris training emails: 7936
no. of Sara training emails: 7884
# **for linear kernal**
1.  training time with 100% data: 25.417 s
2.  training time with 1% data:1.249 s
3.  training time with rbf kernal and 1% data =training time: 1.395 s



In [0]:
accuracy=accuracy_score(labels_test,pred)
print(accuracy)

0.9908987485779295


# **accuracy with kernal='linear'**
1.    acc_whole datasets=0.9840728100113766
2.   acc with 1% dataset =0.8845278725824801
3.   acc with kerenal='rbf'and 1% data =0.6160409556313993
4.   acc with previous kernal and changing value of  c:

   * c=10000 acc=0.8924914675767918
   * c=1000 acc=0.8213879408418657
   * c=100 acc=0.6160409556313993
   * c=10 acc=0.6160409556313993

